### PR TITLE
p2p: Add network simulation framework

### PR DESCRIPTION
**Pull request by lmars**
This PR introduces a network simulation framework which can be used to run simulated networks of devp2p nodes.

The intention is to use this for testing protocols, performing benchmarks and visualising emergent network behaviour.

See [p2p/simulations/README.md](https://github.com/ethersphere/go-ethereum/blob/p2p-simulations/p2p/simulations/README.md) for a detailed description.

This replaces #3102.

### CODE REVIEW

**Comment by zelig** [comment]
This PR is a work of the last 8 months and went through several phases of evolution. The version can be considered pretty extensively tested and carefully reviewed. But I dont want to put me as a reviewer.

File: p2p/simulations/cmd/p2psim/main.go [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+//     $ p2psim node create
+//     Created node02
+//
+//     $ p2psim node start node02
+//     Started node02
+//
+//     $ p2psim node connect node01 node02
+//     Connected node01 to node02
+//
+package main
```
Comment (fjl) [inline p2p/simulations/cmd/p2psim/main.go]: Please move this to the top level cmd directory.

File: p2p/message.go [hunk 2]
```diff
@@ -271,3 +273,67 @@ func ExpectMsg(r MsgReader, code uint64, content interface{}) error {
 	}
 	return nil
 }
+
+// MsgEventer wraps a MsgReadWriter and sends events whenever a message is sent
+// or received
+type MsgEventer struct {
```
Comment (fjl) [inline p2p/message.go]: Please unexport this.

**Comment by lmars** [comment]
@fjl comments addressed in cd12a9e and 30810aa (I have also rebased on master)

**Comment by fjl** [comment]
@lmars please fix tests

**Comment by lmars** [comment]
@fjl I fixed the message filter tests in 6127cfe, one of the Travis builds timed out but I'm not sure if that is an intermittent failure on master as it seems unrelated to the changes in this PR.

**Pull request merged**